src/Utilities/Functional.hpp

```diff
@@ -0,0 +1,255 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+/// \file
+/// Defines function objects
+
+#include <complex>
+
+#pragma once
+
+/*!
+ * \ingroup UtilitiesGroup
+ * \brief creates unary functors from functions
+ *
+ * \details
+ * creates a struct with name STRUCTNAME which calls function FUNC for any
+ * parameter. It can be supplied an explicit type,
+ * e.g. `STRUCTNAME<double>{}(val)`, or can infer the type if no type is
+ * supplied, e.g. `STRUCTNAME<>{}(val)`
+ *
+ */
+#define MAKE_UNARY_STRUCT(STRUCTNAME, FUNC)               \
+  template <typename T = void>                            \
+  struct STRUCTNAME {                                     \
+    auto operator()(const T& arg) { return (FUNC(arg)); } \
+  };                                                      \
+  template <>                                             \
+  struct STRUCTNAME<void> {                               \
+    template <typename T>                                 \
+    auto operator()(const T& arg) {                       \
+      return (FUNC(arg));                                 \
+    }                                                     \
+  };
+
+/*!
+ * \ingroup UtilitiesGroup
+ * \brief creates binary functors from functions
+ *
+ * \details
+ * creates a struct with name STRUCTNAME which calls function FUNC for any
+ * parameter. It can be supplied an explicit type,
+ * e.g. `STRUCTNAME<double,double>{}(val1,val2)`, or can infer the type if no
+ * type is supplied, e.g. `STRUCTNAME<>{}(val1,val2)`
+ *
+ */
+#define MAKE_BINARY_STRUCT(STRUCTNAME, FUNC)          \
+  template <typename T1 = void, typename T2 = void>   \
```

Comment:
This is an interesting point -- I think it is more desirable to have separate template parameters, and perhaps so much so to deviate from the STL. For my purposes in developing general tests, it would be nice to have a functor that works for testing the addition of, for instance, DataVectors and doubles, which is permitted for the raw operator +. If we want to deviate, though, we'd want our own Plus, Minus, etc. as well.